### ISSUE TITLE
Kernel boot hangs when CONFIG_NR_CPUS is higher than 64

### ISSUE THREAD

**Issue by lenticularis39** [linux-ia64/linux-ia64#3]
This was observed also on upstream kernel before IA-64 removal.

**Comment by johnny-mnemonic** [comment]
`arch/ia64/Kconfig` has:
```
[...]
config NR_CPUS
	int "Maximum number of CPUs (2-4096)"
	range 2 4096
	depends on SMP
	default "4096"
	help
	  You should set this to the number of CPUs in your system, but
	  keep in mind that a kernel compiled for, e.g., 2 CPUs will boot but
	  only use 2 CPUs on a >2 CPU system.  Setting this to a value larger
	  than 64 will cause the use of a CPU mask array, causing a small
	  performance hit.
[...]
```
...so the problem could be related to this _"use of a CPU mask array"_.

**Comment by lenticularis39** [comment]
Oh, is there maybe an alignment bug? :thinking:

**Comment by Cyanoxygen** [comment]
The `generic_defconfig` has NR_CPUS set to the maximum value, rendering the kernel unbootable, it just reboots upon execution. I couldn't get `earlycon` to work (I guess it is because of the serial hardware being a PCI adapter), so I don't know what exactly happened.

**Comment by lenticularis39** [comment]
>  it just reboots upon execution

@Cyanoxygen Do your lights turn red before the reboot? In that case, it's a MCE, and the point where it crashed (usually from the return address in `br0`) can be collected from that and matched to the kernel code.

I might have not been patient enough to wait for the MCE to finish collecting, it takes about 15 seconds on my machine. I'll look at this against this week. I'm curious: do you have a use case for this?

**Comment by Cyanoxygen** [comment]
> and the point where it crashed (usually from the return address in `br0`) can be collected from that and matched to the kernel code

Our machine is a rx2660, and the machine owner said that the light indeed flashed red before it resets. I looked through the System Log, and found some `OS_MCA_UNCORRECTED_MC` entries which may match your description:

<img width="1137" height="602" alt="Image" src="https://github.com/user-attachments/assets/53da0981-6ae1-448c-8168-68d7cb45dc36" />

The machine is not mine, however I can access its serial console remotely by SSHing to a machine within its local network and has a physical serial connection into the serial port.

> do you have a use case for this

We are going to revive our effort to bring a modern Linux distribution to old machines (formally "AOSC OS/Retro"), and IA-64 is one of them.

**Comment by lenticularis39** [comment]
@Cyanoxygen Yep, that is an MCE. I collect the data from my machine and will try to fix the bug, no need for you to collect the data on yours :)

> We are going to revive our effort to bring a modern Linux distribution to old machines (formally "AOSC OS/Retro"), and IA-64 is one of them.

Ah I see. We are doing the same thing with [T2 Linux](https://t2sde.org/). I have an rx2620 personally but we also have a rx2660.

**Comment by lenticularis39** [comment]
Hmm, my machine just hung on loading initramfs, no MCE so far: [6.16.6-epic1+-nr-cpus-64-crash-dmesg.log](https://github.com/user-attachments/files/22259941/6.16.6-epic1%2B-nr-cpus-64-crash-dmesg.log)

**Comment by Cyanoxygen** [comment]
You may try this config instead (derived from `generic_defconfig` with 6.16.0-epic1 kernel):
https://repo-hk.aosc.io/misc/kconfig-6.16.0-epic1-aosc-main-nrcpu4096%2B

**Comment by lenticularis39** [comment]
Fixed by 0e41699e3e7231efed9ccf75619c7987de5ad74e.
